cpp/src/io/csv/csv_gpu.h

```diff
@@ -16,13 +16,179 @@
 
 #pragma once
 
+#include <rmm/thrust_rmm_allocator.h>
 #include <cudf/types.hpp>
 #include <io/utilities/parsing_utils.cuh>
 
 namespace cudf {
 namespace io {
 namespace csv {
 namespace gpu {
+
+/**
+ * CSV row parsing context
+ * NONE: No special context (normal parsing)
+ * QUOTE: Within a quoted field
+ * COMMENT: Within a comment line (discard every character until terminator)
+ * EOF: End state (EOF reached)
+ **/
+enum { ROW_CTX_NONE = 0, ROW_CTX_QUOTE = 1, ROW_CTX_COMMENT = 2, ROW_CTX_EOF = 3 };
+
+constexpr uint32_t rowofs_block_dim = 512;
+/// Character block size for gather_row_offsets
+constexpr uint32_t rowofs_block_bytes = rowofs_block_dim * 32;  // 16KB/threadblock
+
+/**
+ * Row parsing context with row count
+ * Format: row_count * 4 + id, where `row_count` is the number of rows
+ * in a character block, and `id` is the row parser state at the end of the block.
+ **/
+typedef uint32_t rowctx32_t;
+typedef uint64_t rowctx64_t;
+
+/**
+ * Packed row context format
+ *
+ * Pack four rowctx32_t values, where each value represents the output row context
+ * for one of four possible input contexts when parsing a character block.
+ * Each rowctx32_t value is truncated to 20-bit (limiting the max number of rows
+ * to 18-bit) and concatenated to form a 80-bit value, whose upper 16 bits are
+ * always zero (EOF input state implies a zero row count) and therefore
+ * stored as 64-bit.
+ **/
+typedef uint64_t packed_rowctx_t;
+
+/**
+ * @brief return a row context from a {count, id} pair
+ *
+ * The 32-bit row context consists of the 2-bit parser state stored in the lower 2-bits
+ * and a 30-bit row count in the upper 30 bits.
+ *
+ **/
+inline __host__ __device__ rowctx32_t make_row_context(uint32_t row_count, uint32_t out_ctx)
+{
+  return (row_count << 2) + out_ctx;
+}
+
+/**
+ * @brief pack multiple row contexts together
+ *
+ * The 64-bit packed row context format represents the four possible output row context state
+ * from each of the four possible input row context state.
+ * Each output state consists of the 2-bit row context state along with a 18-bit row count
+ * value (row count is assumed to be a local count that fits in 18-bit)
+ * The four 20-bit values are concatenated to form a 80-bit value, truncated to 64-bit
+ * since a block starting in a EOF state can only have a zero row count (and the output
+ * state corresponding to an EOF input state can only be EOF, so only the first 3 output
+ * states are included as parameters, and the EOF->EOF state transition is hardcoded)
+ *
+ **/
+inline __host__ __device__ packed_rowctx_t pack_row_contexts(rowctx32_t ctx0,
```

Comment:
Regardless of what the sales brochure says, `constrexpr` is largely cosmetic (the problem of reducing constant expressions at compile time was solved decades ago). AFAICT, the primary use of constexpr is for compile-time asserts, though it may act as a sort of encouragement for the compiler for complex functions (such as loops/recursivity) where it may have given up attempting compile-time evaluation (clearly not the case here)
I can make it constexpr, though (why not, I suppose, it makes zero difference)